# Pipeline_ATACSeq

## *** UNDER CONSTRUCTION ***

Correspondence = jackfreeland01@gmail.com 

### Steps 

- [Quality Control on Raw Reads](#Quality-Control-on-Raw-Reads)
- [Adapter Trimming](#Adapter-Trimming)
- [Quality Control on Trimmed Reads](#Quality-Control-on-Trimmed-Reads)
- [Alignment](#Alignment)

## Quality Control on Raw Reads
Before initiating formal processing steps, it is advisable to assess the overall quality of the raw sequencing files. This pipeline uses FASTQC to generate [FastQC reports](https://dnacore.missouri.edu/PDF/FastQC_Manual.pdf) to provide information on sequence quality, GC content, length distribution, duplicate sequences, overrepresented sequences, K-mer content and adapter contamination. 

If interested, some alternatives to [FASTQC](#https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) are [FastP](#) (doi: 10.1093/bioinformatics/bty560) or [BBDuk](#).


## Adaptor Trimming
## Quality Control on Trimmed Reads
## Alignment
